shared/src/balancer_event_handler.rs

```diff
@@ -54,15 +117,127 @@ impl PoolSpecialization {
             0 => Ok(Self::General),
             1 => Ok(Self::MinimalSwapInfo),
             2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {}", t)),
+            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
         }
     }
 }
 
-#[derive(Default)]
-pub struct BalancerPools {}
-
 impl BalancerPools {
+    fn try_upgrade(&mut self, pool_builder: WeightedPoolBuilder) {
+        match pool_builder.clone().try_into() {
+            Ok(weighted_pool) => {
+                let token_registration = pool_builder.tokens_registration.expect("known to exist");
+                // When upgradable, delete pending pool and add to valid pools
+                let pool_id = token_registration.pool_id;
+                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
+                self.pending_pools.remove(&pool_id);
+                self.pools.insert(pool_id, weighted_pool);
+                for token in token_registration.tokens {
+                    self.pools_by_token
+                        .entry(token)
+                        .or_default()
+                        .insert(pool_id);
+                }
+            }
+            Err(_) => {
+                tracing::info!("Pool Builder not yet upgradable");
+            }
+        }
+    }
+
+    // All insertions happen in one transaction.
+    fn insert_events(&mut self, events: Vec<(EventIndex, BalancerEvent)>) -> Result<()> {
+        for (index, event) in events {
+            let pool_builder = match event {
+                BalancerEvent::PoolRegistered(event) => self.insert_pool(index, event),
+                BalancerEvent::TokensRegistered(event) => self.insert_token_data(index, event),
+            };
+            // In the future, when processing TokensDeregistered we may have to downgrade the result.
+            self.try_upgrade(pool_builder)
+        }
+        Ok(())
+    }
+
+    fn insert_pool(
```

Comment:
It's weird that this method mutates self (by adding the pool builder) and also returns a copy of the builder. Why is that? If for testing purposes, could we inspect self.pending_pools in the test directly? I think this method should return `()`

Reply:
Yes. I mentioned this in my most recent comment on the PR. Will change.